### PR TITLE
Allow rescale_mode=None in plot to only clip

### LINKED ISSUE #1280: `plot` should allow no rescaling whatsoever
Let's say you have a scientific image in 0-1, and want to plot it without any rescaling i.e. 0 = black, 1 = white, and anything in between is grey.

Currently in `deepinv.utils.plot` we don't have that option. Even when `rescale_mode=clip` is used, it's clipped to 0-1, but `ax.imshow` is still performing rescaling such that black = min value of data rather than 0.

The current work around is to use `imshow_kwargs`:

```python
from matplotlib.colors import Normalize
norm = Normalize(0, 1, clip=True)
plot(...,  norm=norm, rescale_mode="clip")
```
We should probably have a None option for rescale_mode which does this. Side question is how to handle the clip arg of Normalize.

### PR DESCRIPTION
**Pull request by dldou**
Fixes #1280.

This PR:
- adds `rescale_mode=None` to `deepinv.utils.plot`, keeping image values unchanged while displaying them over `[0, 1]` with `Normalize(0, 1, clip=True)`;
- preserves the existing `vmin`/`vmax` warning behavior;
- applies the same normalization to plot insets.

Tests cover the existing `"clip"` behavior and the new `None` mode.

Thank you for contributing to DeepInverse!

Please read our [contributing guidelines](https://deepinv.org/contributing.html) for instructions on how to contribute, including writing tests, documentation and code style.

Once the GitHub tests have been approved by a maintainer (only required for first-time contributors), and the `Build Docs` GitHub action has run successfully, you can download the documentation as a zip file from the [Actions page](https://github.com/deepinv/deepinv/actions/workflows/docs_cpu.yml). Look for the workflow run corresponding to your pull request.

### Checks to be done before submitting your PR

- [ ] `python3 -m pytest deepinv/tests` runs successfully.
- [X] `black .` and `ruff check .` run successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [X] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [changelog.rst](https://github.com/deepinv/deepinv/blob/main/docs/source/changelog.rst).

### LLM policy

LLM usage is ok, but not PRs generated 100% by AI. See our [LLM policy](https://deepinv.org/contributing.html#llm-policy) Tick below as appropriate:

- [ ] I did not use LLM tools to write the code
- [X] I, a human, wrote code, and LLM tools helped me to write part of the code.
- [ ] An LLM tool wrote all of the code.
- [ ] An agent submitted the PR and wrote the description.

### Notes on tests and make html
The local documentation build could not complete because of the MPS backend. Some operations  are unsupported by the MPS backend (for instance ,`float64`, `torch.poisson`, and `nan_to_num` on complex tensors). 

Some tests of the suite failed because of my MPS backend but the 8 targeted test cases for this PR all pass:
```
python -m pytest \
  deepinv/tests/test_utils.py::test_plot_clip_rescale_mode \
  deepinv/tests/test_utils.py::test_plot_rescale_mode_none \
  deepinv/tests/test_utils.py::test_plot_rescale_mode_none_vmin_vmax_warning \
  deepinv/tests/test_utils.py::test_plot_rescale_mode_none_inset
```

### CODE REVIEW

**Comment by Tmodrzyk** [comment]
Hi @dldou thank you so much for fixing this. Don't hesitate to reach out if you need another workflow run or a review ! :smile:

**Comment by dldou** [comment]
Hi @Tmodrzyk, thanks a lot! I added a small test update to cover the missing lines reported by codecov (the colorbar lines weren't tested). The workflows need approval again. Thanks in advance! 😉

**Comment by dldou** [comment]
@Tmodrzyk, everything looks good on my side now. The PR should be ready for review whenever you have time. Thanks again for your help!

File: deepinv/tests/test_utils.py [hunk 1]
```diff
@@ -279,6 +279,92 @@ def test_plot(
             assert axs is None
 
 
+@pytest.mark.parametrize(
+    "vmin, vmax, expected_clim",
```
Comment (Tmodrzyk) [inline deepinv/tests/test_utils.py]:
```suggestion
    "vmin, vmax, expected_clip",
```

File: deepinv/tests/test_utils.py [hunk 2]
```diff
@@ -279,6 +279,92 @@ def test_plot(
...
+    assert plotted_image.get_clim() == pytest.approx(expected_clim, abs=1e-5)
+
+
+@pytest.mark.parametrize(
+    "image",
+    [
+        torch.tensor([[[0.25, 0.5], [0.5, 1.25]]]),
+        torch.tensor([[[-0.25, 0.5], [0.5, 0.75]]]),
+    ],
+)
```
Comment (Tmodrzyk) [inline deepinv/tests/test_utils.py]: We can put this inside the body of the test, and rather parametrize other parameter of the plotting function such as `cbar`, `return_axs` etc.

File: deepinv/tests/test_utils.py [hunk 3]
```diff
@@ -279,6 +279,92 @@ def test_plot(
...
+        image,
+        rescale_mode=None,
+        plot_inset=True,
+        show=False,
+        return_axs=True,
+    )
+    inset_image = axs[0, 0].child_axes[0].images[0]
+
+    assert inset_image.get_clim() == pytest.approx((0.0, 1.0), abs=1e-5)
+    assert inset_image.norm.clip
```
Comment (Tmodrzyk) [inline deepinv/tests/test_utils.py]: Overall, all these new tests should be factored into one, or even better, integrated in `test_plot` 😄

File: deepinv/utils/plotting.py [hunk 4]
```diff
@@ -448,10 +463,16 @@ def plot(
         plt.suptitle(suptitle, wrap=True)
         fig.subplots_adjust(top=0.75)
 
+    if rescale_mode is None:
+        imshow_kwargs.setdefault("norm", Normalize(0.0, 1.0, clip=True))
```
Comment (Tmodrzyk) [inline deepinv/utils/plotting.py]: Since we're only passing the "no normalization" information to `imshow`, are we sure that the saved image will have the same range as the displayed one ?

**Review by Tmodrzyk — DISMISSED** [review]

**Comment by Tmodrzyk** [comment]
Hi @dldou thank you again for this PR. The fix you propose seems to work !
My main request is to factor the tests, and if possible check that saved images match the displayed one, I'm not sure that it is the case at the moment. :smile:

File: deepinv/utils/plotting.py [hunk 5]
```diff
@@ -202,16 +202,17 @@ def preprocess_img(
...
 
     .. note::
 
         Real-valued tensors with two channels are assumed to be Cartesian
         representation of complex images and are processed accordingly.
 
     :param torch.Tensor im: the batch of images to preprocess, it is expected to be of shape (B, C, *).
-    :param str rescale_mode: the normalization mode, either 'min_max' or 'clip'.
+    :param str, None rescale_mode: the normalization mode, either ``'min_max'``, ``'clip'``, or ``None``.
+        If ``None``, the image values are not rescaled.
```
Comment (Andrewwango) [inline deepinv/utils/plotting.py]: This is a bit vague, pixel values are always plotted to some scale. min_max -> scale is such that white is max pixel value. clip -> scale is such that white is 1, but anything less than 1 is still rescaled (this is the original bug from the issue, we need to explain that here). None -> scale is such that white is 1.

File: deepinv/utils/plotting.py [hunk 6]
```diff
@@ -242,13 +243,25 @@ def preprocess_img(
             if not isinstance(mins, list):
                 mins, maxs = [mins], [maxs]
             scales = list(zip(mins, maxs, strict=True))
-        else:  # clip
+        elif rescale_mode == "clip":
             v0 = vmin if vmin is not None else 0.0
             v1 = vmax if vmax is not None else 1.0
             scales = [(v0, v1)] * im.shape[0]
+        else:  # rescale_mode is None
+            v0 = 0.0
```
Comment (Andrewwango) [inline deepinv/utils/plotting.py]: why can't we still allow vmin, vmax to be used to rescale_mode=None?

File: deepinv/utils/plotting.py (hunk 4 above)
Comment (Andrewwango) [inline deepinv/utils/plotting.py]: kwargs setdefault is difficult to understand for users and could lead to some confusing behaviour. I would prefer an explicit if else

**Review by Andrewwango — CHANGES_REQUESTED** [review]
Thanks for contributing this fix. Please refer to the comments and the original issue on how we should be documenting this fix

**Comment by dldou** [comment]
Thank you for your feedbacks! I made a few small changes to address the issues you found: 
- `rescale_mode=None` now supports `vmin`/`vmax` for clip-only-clip and no rescaling
- the same normalization is consistently used by `imshow` and saved images now use `mpl_img.write_png()` which also remove the need for `np.ascontiguousarray`
- all the small function tests are now integrated into ``test_plot``function
- the documentation for ``rescale_mode`` in ``preprocess_img`` and ``plot`` has been clarified

File: docs/source/changelog.rst [hunk 7]
```diff
@@ -36,6 +36,7 @@ New Features
 - Add :class:`deepinv.loss.metric.BRISQUE` and :class:`deepinv.loss.metric.NIMA` no-reference image quality metrics (:gh:`1310` by `Julian Tachella`_)
 - Add :class:`deepinv.datasets.BrainWebPET` (:gh:`1286` by `Thibaut Modrzyk`_)
 - Add ``use_dict_output`` option to every dataset class, returning a dict ``{"x", "y", "params"}`` instead of a tuple; propagate support to all deepinv internals (:gh:`1244` by `Romain Vo`_)
+- Add :func:`deepinv.utils.plot` to disable image rescaling with ``rescale_mode=None``. (:gh:`1339` by dldou)
```
Comment (Andrewwango) [inline docs/source/changelog.rst]: if you want, you can attribute your name and include a link below, so that people can see who you are!

You need to add a line for yourself at the bottom e.g.

```
.. _Julian Tachella: https://github.com/tachella
```

then when you do



```suggestion
- Add :func:`deepinv.utils.plot` to disable image rescaling with ``rescale_mode=None``. (:gh:`1339` by`Delphine Doutsas`_)
```

it appears with the link as

Image

File: deepinv/utils/plotting.py [hunk 8]
```diff
@@ -202,18 +202,19 @@ def preprocess_img(
...
         Real-valued tensors with two channels are assumed to be Cartesian
         representation of complex images and are processed accordingly.
 
     :param torch.Tensor im: the batch of images to preprocess, it is expected to be of shape (B, C, *).
-    :param str rescale_mode: the normalization mode, either 'min_max' or 'clip'.
-    :param float, None vmin: minimum value for clipping when using 'clip' rescaling.
-    :param float, None vmax: maximum value for clipping when using 'clip' rescaling.
+    :param str, None rescale_mode: the normalization mode, either ``'min_max'``, ``'clip'``, or ``None``.
+        With ``None``, image values are clipped to ``[vmin, vmax]`` without being rescaled.
+    :param float, None vmin: minimum clipping bound when using ``'clip'`` or ``None``. Defaults to 0.
```
Comment (Andrewwango) [inline deepinv/utils/plotting.py:216]:
```suggestion
    :param float, None vmin: minimum clipping bound when using `rescale_mode=`'clip'`` or ``rescale_mode=None``. Defaults to 0.
```

File: deepinv/utils/plotting.py [hunk 9]
```diff
@@ -202,18 +202,19 @@ def preprocess_img(
...
+    :param float, None vmax: maximum clipping bound when using ``'clip'`` or ``None``. Defaults to 1.
```
Comment (Andrewwango) [inline deepinv/utils/plotting.py:217]:
```suggestion
    :param float, None vmax: maximum clipping bound when using ``rescale_mode='clip'`` or ``rescale_mode=None``. Defaults to 1.
```

File: deepinv/utils/plotting.py [hunk 10]
```diff
@@ -448,20 +457,32 @@ def plot(
         plt.suptitle(suptitle, wrap=True)
         fig.subplots_adjust(top=0.75)
 
+    norm = imshow_kwargs.pop("norm", None)
+
+    if rescale_mode is None and norm is None:
+        norm = Normalize(0.0, 1.0, clip=True)
```
Comment (Andrewwango) [inline deepinv/utils/plotting.py:465]: Do you think this should use vmin, vmax instead of 0, 1?

This way, when rescale_mode=None, vmax = "white" and vmin = "black". Currently, this always assumes white=1, even if vmin>1.

Comment (Andrewwango) [reply deepinv/utils/plotting.py:465]: Image

**Review by Andrewwango — CHANGES_REQUESTED** [review]
Thanks @dldou ! I have some minor comments + one question.

File: deepinv/utils/plotting.py (hunk 10 above)
Comment (dldou) [reply deepinv/utils/plotting.py:465]: Since `rescale_mode=None` can be combined with explicit `vmin` and/or `vmax`, users can also pass a custom `norm` to control how the clipped values are displayed. So I'd prefer to keep the current behavior unchanged, if that works for you.

**Review by Andrewwango — APPROVED** [review]

File: docs/source/changelog.rst [hunk 11]
```diff
@@ -67,9 +68,7 @@ Fixed
 - (Breaking) Have `x_shift` represent horizontal shifts and `y_shift` vertical shifts in :class:`deepinv.transform.Shift` (:gh:`1236` by `Sarra Amiri`_)
 - Force trainer non_blocking_transfers=False on MPS and CPU (:gh:`1311` by `Andrew Wang`_)
 - Fix :class:`deepinv.physics.PET` incorrect device attribution of attenuation and background on update and incorrect handling of batched attenuation  (:gh:`1331` by `Thibaut Modrzyk`_)
-
-
-
+- Fix the channel description in NBUDataset documentation (:gh:`` by dldou)
```
Comment (jscanvic) [inline docs/source/changelog.rst]: Looks like a failed merge conflict

**Review by jscanvic — COMMENTED** [review]
Thanks a lot for fixing this very non-trivial issue Delphine!

Ty Andrew & Thibaut for your reviews as well

I fixed very minor things
- In the changelog, I removed a duplicated line, switched the username for the real name, moved the line to the "Current" section instead of v0.4.2
- I added error handling in a if-elif that benefited from a else

I'll approve and merge once CI passes

**Review by jscanvic — APPROVED** [review]

**Comment by dldou** [comment]
Thank you everybody! :+1:
